feat: Add `filterEntrypoints` option to speed up development

yunsii opened the pull request:
ref: https://github.com/wxt-dev/wxt/issues/330

Add a simple test case and passed, but I found that test cases of `src/core/utils/building/__tests__/find-entrypoints.test.ts` seems outdated, not only caused by my changes:

![image](https://github.com/wxt-dev/wxt/assets/18096089/00c9c936-dcce-4a43-9950-2ba786281e15)

aklinker1 commented on src/types/external.ts:
```diff
@@ -31,6 +31,12 @@ export interface InlineConfig {
+   *
+   * @default true
+   */
+  allowEntrypoints?: true | string[];
```
I'd prefer just `string[]` here. True is the same as undefined, I don't want more than one type to have the same meaning.

```suggestion
  allowEntrypoints?: string[];
```

aklinker1 commented on src/types/internal.ts:
```diff
@@ -21,6 +21,7 @@ export interface InternalConfig {
   wxtDir: string;
   typesDir: string;
   entrypointsDir: string;
+  allowEntrypoints: true | string[];
```
```suggestion
  allowEntrypoints: string[] | undefined;
```

aklinker1 commented on src/core/utils/building/find-entrypoints.ts:
```diff
@@ -125,6 +144,9 @@ export async function findEntrypoints(
     if (include?.length && !exclude?.length) {
       return include.includes(config.browser);
     }
+    if (disallowedEntrypointNames.includes(entry.name)) {
+      return false;
+    }
```
Now that I think about it, putting this inside `findEntrypoints` is wrong. When the `.wxt` directory is updated, the `.wxt/types/paths.d.ts` will also be filtered. We should probably move the filter to the `rebuild` function to prevent that from happening...

But we also need to move the `include` and `exclude` checks to `rebuild` for the same reason.

Let's keep this as-is for the this PR, no changes required. Let's move the filtering to `rebuild` in a separate, future PR.

yunsii replied:
> But we also need to move the `include` and `exclude` checks to `rebuild` for the same reason.

Yes, I also found this problem.

aklinker1 reviewed (commented):
Looks good! I think I prefer `filterEntrypoints` now that I see it implemented. So rename:

- `allowEntrypoints` &rarr; `filterEntrypoints`
- `-e, --allow-entrypoint <entrypoint>` &rarr; `-e, --filter-entrypoint <entrypoint>`
- `disallowed` &rarr; `skipped`

Once the PR is opened for review, I'll go in and update the wording of some of the comments.

aklinker1 commented:
> Add a simple test case and passed, but I found that test cases of src/core/utils/building/__tests__/find-entrypoints.test.ts seems outdated, not only caused by my changes:

They are not out of date, must be something with your setup. CI ensures those tests are passing on both Windows and Linux. I use mac for development, and can confirm they work on Mac as well.

https://github.com/wxt-dev/wxt/actions/workflows/validate.yml

yunsii commented:
I use Ubuntu in WSL, and test cases failed like following:

```diff
 FAIL  |unit| src/core/utils/building/__tests__/find-entrypoints.test.ts > findEntrypoints > should find entrypoint for content.css
 FAIL  |unit| src/core/utils/building/__tests__/find-entrypoints.test.ts > findEntrypoints > should find entrypoint for content/index.css
AssertionError: expected { name: 'content', …(5) } to deeply equal { type: 'content-script-style', …(4) }

- Expected
+ Received

  Object {
    "inputPath": "/src/entrypoints/content.css",
    "name": "content",
-   "options": Object {},
+   "options": Object {
+     "exclude": undefined,
+     "include": undefined,
+   },
    "outputDir": "/home/yuns/github/wxt/.output/content-scripts",
+   "skipped": false,
    "type": "content-script-style",
  }

 ❯ src/core/utils/building/__tests__/find-entrypoints.test.ts:561:28
    559| 
    560|     expect(entrypoints).toHaveLength(1);
    561|     expect(entrypoints[0]).toEqual(expected);
       |                            ^
    562|   });
    563| 
```

I only add `skipped` field, the `options` field change is strange.

yunsii commented:
> I use Ubuntu in WSL, and test cases failed like following:
> 
> ```diff
>  FAIL  |unit| src/core/utils/building/__tests__/find-entrypoints.test.ts > findEntrypoints > should find entrypoint for content.css
>  FAIL  |unit| src/core/utils/building/__tests__/find-entrypoints.test.ts > findEntrypoints > should find entrypoint for content/index.css
> AssertionError: expected { name: 'content', …(5) } to deeply equal { type: 'content-script-style', …(4) }
> 
> - Expected
> + Received
> 
>   Object {
>     "inputPath": "/src/entrypoints/content.css",
>     "name": "content",
> -   "options": Object {},
> +   "options": Object {
> +     "exclude": undefined,
> +     "include": undefined,
> +   },
>     "outputDir": "/home/yuns/github/wxt/.output/content-scripts",
> +   "skipped": false,
>     "type": "content-script-style",
>   }
> 
>  ❯ src/core/utils/building/__tests__/find-entrypoints.test.ts:561:28
>     559| 
>     560|     expect(entrypoints).toHaveLength(1);
>     561|     expect(entrypoints[0]).toEqual(expected);
>        |                            ^
>     562|   });
>     563| 
> ```
> 
> I add `skipped` field only, the `options` field change is strange.

Got it, the root cause is that I introduced the `skipped` field, if I fixed it, the diff of `options` also disappeared 😂

yunsii commented:
Any progress? we really need this feature.

aklinker1 reviewed (approved):
CLI wasn't working, I fixed that. Also added colors to the list of filtered entrypoints:

<img width="523" alt="Screenshot 2024-01-16 at 1 50 21 PM" src="https://github.com/wxt-dev/wxt/assets/10101283/aeba1ab2-b9fe-475e-8f6f-5110038023f0">

yunsii commented:
> CLI wasn't working, I fixed that. Also added colors to the list of filtered entrypoints:
> 
> <img width="523" alt="Screenshot 2024-01-16 at 1 50 21 PM" src="https://github.com/wxt-dev/wxt/assets/10101283/aeba1ab2-b9fe-475e-8f6f-5110038023f0">
> 

Is it possible to add test cases for CLI?

aklinker1 commented:
Probably... I've never done it before though. Something like this might work.

```ts
vi.mock("../build");
const buildMock = vi.mocked(build)

describe("CLI", () => {
  beforeEach(() => {
    vi.resetModules()
  });

  it("should ...", () => {
    process.argv = ["wxt", "build", ...];
    await import("../CLI");

    expect(buildMock).toBeCalledWith(...)
  });
})
```

Shouldn't need E2E tests just to check if the flags are parsed correctly, unit tests should be enough.

yunsii commented on src/cli.ts:
```diff
@@ -165,3 +181,12 @@ function wrapAction(
+ * Array flags, when not passed, are either `undefined` or `[undefined]`. This function filters out
+ * the
+ */
+function getArrayFromFlags<T>(flags: any, name: string): T[] | undefined {
+  const array = [flags[name]].flat() as Array<T | undefined>;
+  return array.filter((item) => item != null) as T[];
+}
```
@aklinker1 The result will be empty array at least, so it caused `userConfig` of `mergeInlineConfig(inlineConfig, userConfig)` has no effect.

After released v1.4.2, I have to patch wxt temporarily.
